Use mio to replace Epoll

uran0sH opened the pull request:
Epoll is linux-specific. So we use mio, which is a cross-platform event notification, to replace Epoll.
This is a draft version.

uran0sH committed:
Use mio to replace Epoll

Epoll is linux-specific. So we use mio, which is a cross-platform
event notification, to replace Epoll.

Signed-off-by: Wenyu Huang <huangwenyuu@outlook.com>

uran0sH commented:
cc@stefano-garzarella @germag @osteffenrh

uran0sH commented on vhost-user-backend/src/backend.rs:
```diff
@@ -144,7 +146,7 @@ pub trait VhostUserBackend: Send + Sync {
         &self,
         device_event: u16,
-        evset: EventSet,
+        evset: VringEventSet,
```
Yeah, I think we can use mio::Interest. Originally, since `mio::Event` can't convert to Interest, I converted the `mio::Event` to `VringEventSet`. Now I think we can implement a method to directly convert Event to Interest

stefano-garzarella replied:
Yeah, if we can use `mio` stuff, it will be great IMO, also if we need to refactor the code to avoid the conversion and use `mio:Event` directly

uran0sH commented on vhost-user-backend/src/event_loop.rs:
```diff
@@ -134,14 +205,62 @@ where
+                Token(data as usize),
+                Interest::from(ev_type),
+            )
+            .map_err(|e| io::Error::new(io::ErrorKind::Other, e))
```
Need to fix it. After error, it also needs to remove (interest, data) from map.

uran0sH commented:
I found a deadlock problem. https://github.com/uran0sH/vhost/pull/2/files#diff-e1c9dc7802234bc118de8314ec9cacc98066eea325a6f14ff5887ece3466d5a3R260 Here will hold the lock to listen for events's arrival. https://github.com/uran0sH/vhost/pull/2/files#diff-e1c9dc7802234bc118de8314ec9cacc98066eea325a6f14ff5887ece3466d5a3R193 A lock is also needed here to register events. So this is what happens
```
      thread A                                          thread B
poller.lock().unwrap().poll()
         |
         | hold the lock            poller.lock().unwrap().registry().register()
         |
```
solve this problem by using the following method, but is there any other better method? @stefano-garzarella @germag @osteffenrh
```
        'poll: loop {
            self.poller
                .write()
                .unwrap()
                .poll(&mut events, Some(Duration::from_secs(1)))
                .map_err(VringEpollError::EpollWait)?;

            thread::yield_now();
        }
```

stefano-garzarella commented:
@uran0sH looking at the documentation, it seems that `Registry` has a `try_clone()` method: https://docs.rs/mio/latest/mio/struct.Registry.html#method.try_clone

So, when you create `self.poller`, you can clone the `registry`. In this way in `register_event` you can use the clone, instead of holding the mutex of the event loop.

WDYT?

germag commented on vhost-user-backend/src/event_loop.rs:
```diff
@@ -7,57 +7,107 @@ use std::fmt::{Display, Formatter};
+            .map(|interest| Interest::WRITABLE | interest)
+            .or(Some(Interest::WRITABLE));
+    }
+    #[cfg(any(
+        target_os = "dragonfly",
+        target_os = "freebsd",
+        target_os = "ios",
+        target_os = "macos",
+        target_os = "tvos",
+        target_os = "visionos",
+        target_os = "watchos",
+    ))]
+    if value.is_aio() {
+        interest = interest
+            .map(|interest| Interest::AIO | interest)
+            .or(Some(Interest::AIO));
+    }
+    #[cfg(target_os = "freebsd")]
+    if value.is_lio() {
+        interest = interest
+            .map(|interest| Interest::LIO | interest)
+            .or(Some(Interest::LIO));
+    }
+    #[cfg(any(target_os = "linux", target_os = "android"))]
+    if value.is_priority() {
+        interest = interest
+            .map(|interest| Interest::PRIORITY | interest)
+            .or(Some(Interest::PRIORITY));
+    }
```
please explain why this `#[cfg(...)]` are required

uran0sH commented on vhost-user-backend/src/event_loop.rs:
```diff
@@ -7,57 +7,107 @@ use std::fmt::{Display, Formatter};
-
+pub type VringEpollResult<T> = std::result::Result<T, VringPollError>;
+
+fn convert_event_to_interest(value: &Event) -> Option<Interest> {
```
https://github.com/tokio-rs/mio/blob/master/src/interest.rs#L35 Get from here.

stefano-garzarella commented on vhost-user-backend/src/event_loop.rs:
```diff
@@ -3,61 +3,113 @@
-
+pub type VringEpollResult<T> = std::result::Result<T, VringPollError>;
+
+// According https://github.com/tokio-rs/mio/blob/master/src/interest.rs#L27
```
Please use a permalink, like https://github.com/tokio-rs/mio/blob/cd972977e7a8234e62c7ba8bd4d4ec1da208e576/src/interest.rs#L27

stefano-garzarella commented on vhost-user-backend/src/event_loop.rs:
```diff
@@ -3,61 +3,113 @@
-    epoll: Epoll,
+    poller: Mutex<Poll>,
+    registry: Registry,
+    // Record which fd and interest are registered
```
Please add "why" we need to register them.

stefano-garzarella commented on vhost-user-backend/src/event_loop.rs:
```diff
@@ -84,22 +136,36 @@ where
         thread_id: usize,
     ) -> VringEpollResult<Self> {
-        let epoll = Epoll::new().map_err(VringEpollError::EpollCreateFd)?;
+        // let epoll = Epoll::new().map_err(VringEpollError::EpollCreateFd)?;
```
leftover

stefano-garzarella commented on vhost-user-backend/src/event_loop.rs:
```diff
@@ -134,14 +200,56 @@ where
+                .iter()
+                .position(|(interest, d)| *interest == ev_type && data == *d)
+            {
+                interests.remove(pos);
```
Should we also call `self.registry.deregister()` and call again `self.registry.register()` with the remaining interestes ?

stefano-garzarella commented on vhost-user-backend/src/event_loop.rs:
```diff
@@ -134,14 +200,56 @@ where
-        self.epoll
-            .ctl(ControlOperation::Add, fd, EpollEvent::new(ev_type, data))
+    pub(crate) fn register_event(&self, fd: RawFd, ev_type: Interest, data: u64) -> Result<()> {
+        self.registry
+            .register(&mut SourceFd(&fd), Token(data as usize), ev_type)
+            .map_err(|e| io::Error::new(io::ErrorKind::Other, e))?;
```
From https://docs.rs/mio/latest/mio/struct.Registry.html#method.register:

> Callers must ensure that if a source being registered with a Poll instance was previously registered with that Poll instance, then a call to [deregister](https://docs.rs/mio/latest/mio/struct.Registry.html#method.deregister) has already occurred. Consecutive calls to register is unspecified behavior.

So maybe we can't do that, without calling `deregister`

stefano-garzarella replied:
Maybe we need to call `reregister`, see https://docs.rs/mio/latest/mio/struct.Registry.html#method.reregister

> Re-registering an event source allows changing the details of the registration. Specifically, it allows updating the associated token and interests specified in previous register and reregister calls.

germag commented:
Today, talking with @stefano-garzarella we agreed that we don't need to use `reregister()`. We only call register once, otherwise `epoll_ctl(..., EPOLL_CTL_ADD, ...)` will fail with `EEXIST` (we also test that `test_vring_epoll_handler()`).
So, we can also assume when we unregister an FD we do that for all its interests.

So, we don't really need the map, since we don't need to track the registered interest, but since the documentation for `mio::Registry::register()` says:
```
Consecutive calls to register is unspecified behavior.
```

We need to keep track of the already register FD's, so a set should be enough. Something that should make `mio::Registry::register()` to be marked as unsafe (and so maybe `VringEpollHandler::register_listener()`) IMO.

I'll look into alternatives (for a more sane API), since on epoll registering twice returns an error instead of UB, and kevent will modify the already registered one.

uran0sH commented:
> Today, talking with @stefano-garzarella we agreed that we don't need to use `reregister()`. We only call register once, otherwise `epoll_ctl(..., EPOLL_CTL_ADD, ...)` will fail with `EEXIST` (we also test that `test_vring_epoll_handler()`). So, we can also assume when we unregister an FD we do that for all its interests.
> 
> So, we don't really need the map, since we don't need to track the registered interest, but since the documentation for `mio::Registry::register()` says:
> 
> ```
> Consecutive calls to register is unspecified behavior.
> ```
> 
> We need to keep track of the already register FD's, so a set should be enough. Something that should make `mio::Registry::register()` to be marked as unsafe (and so maybe `VringEpollHandler::register_listener()`) IMO.
> 
> I'll look into alternatives (for a more sane API), since on epoll registering twice returns an error instead of UB, and kevent will modify the already registered one.

OK. If I understand correctly, normally we will only register each fd once, if it is registered twice we should throw an exception to the user to tell it that it is registered twice. So when calling deregister we can deregister the entire fd

dependabot[bot] committed:
build(deps): bump rust-vmm-ci from `af54708` to `0b1cb86`

Bumps [rust-vmm-ci](https://github.com/rust-vmm/rust-vmm-ci) from `af54708` to `0b1cb86`.
- [Commits](https://github.com/rust-vmm/rust-vmm-ci/compare/af547082938b44b69b6f1bb44c1af8b405d7650c...0b1cb86353cc093f2e17d7e9f6820de80a6c274d)

---
updated-dependencies:
- dependency-name: rust-vmm-ci
  dependency-version: 0b1cb86353cc093f2e17d7e9f6820de80a6c274d
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>

germag commented on vhost-user-backend/src/event_loop.rs:
```diff
@@ -57,6 +58,44 @@ impl std::error::Error for VringPollError {}
+    }
+}
+
+impl From<VringPollEvent> for Interest {
+    fn from(value: VringPollEvent) -> Self {
+        let mut interest = None;
+        if value == VringPollEvent::READABLE {
+            interest = interest
+                .map(|interest| Interest::READABLE | interest)
+                .or(Some(Interest::READABLE));
+        }
+        if value == VringPollEvent::WRITABLE {
+            interest = interest
+                .map(|interest| Interest::WRITABLE | interest)
+                .or(Some(Interest::WRITABLE));
+        }
+        return interest.expect("Interest is invalid");
+    }
+}
```
Where do we use this?, I mean, why we need to convert from `mio::Interest` to `VringPollEvent`?, if we don't do it, better removing it, remember we want to hide `mio::Interest`

uran0sH replied:
We need this one. This one is to convert `VringPollEvent` to `Interest`. It will be used here: https://github.com/uran0sH/vhost/pull/2/files#diff-e1c9dc7802234bc118de8314ec9cacc98066eea325a6f14ff5887ece3466d5a3R242. User can use like this: https://github.com/uran0sH/vhost/pull/2/files#diff-e1c9dc7802234bc118de8314ec9cacc98066eea325a6f14ff5887ece3466d5a3R358

germag replied:
Forget what I said, I see that `mio::Interest` is not public anymore, implementing `impl From<VringPollEvent> for Interest ` is Ok

germag commented on vhost-user-backend/src/event_loop.rs:
```diff
@@ -57,6 +58,44 @@ impl std::error::Error for VringPollError {}
+    }
+}
+
+impl From<Interest> for VringPollEvent {
+    fn from(value: Interest) -> Self {
+        let mut event = VringPollEvent::empty();
+        if value.is_readable() {
+            event |= VringPollEvent::READABLE;
+        }
+        if value.is_writable() {
+            event |= VringPollEvent::WRITABLE;
+        }
+        return event;
+    }
+}
```
doesn't this expose the `mio::Interest` type?, since `VringPollEvent` is public, this impl is also public. Maybe just a private function is enough

uran0sH replied:
This one is unecessary. I will delete it.

germag commented on vhost-user-backend/src/event_loop.rs:
```diff
@@ -57,6 +58,44 @@ impl std::error::Error for VringPollError {}
 
+bitflags! {
+    #[derive(Debug, PartialEq, PartialOrd)]
+    pub struct VringPollEvent: u32 {
```
`VringPollEvent` is too verbose, is a type that will be part of the public API of `VringEpollHandler`, so maybe something like:
- `EventType`
- `Interest` (if you choose this one, please also change the parameters' name `ev_type`  in `register_listener()` and `unregister_listener()`)

are better

germag reviewed (commented):
Hi,

I think is in a good shape, just apply my suggestions and send a PR, we are approaching the end of the gsoc, so we need to have the public PR

uran0sH commented:
See https://github.com/rust-vmm/vhost/pull/316
